velox/connectors/clp/search_lib/ir/ClpIrCursor.cpp

```diff
@@ -0,0 +1,208 @@
+    return ErrorCode::InternalError;
+  }
+  irDeserializer_ = std::make_shared<
+      ::clp::ffi::ir_stream::Deserializer<ClpIrUnitHandler, QueryHandlerType>>(
+      std::move(deserializerResult).value());
+
+  currentSplitLoaded_ = true;
+  return ErrorCode::Success;
+}
+
+std::vector<std::pair<std::string, search::ast::literal_type_bitmask_t>>
+ClpIrCursor::splitFieldsToNamesAndTypes() const {
+  auto result = std::vector<
+      std::pair<std::string, search::ast::literal_type_bitmask_t>>{};
+  for (size_t i{0}; i < outputColumns_.size(); ++i) {
+    auto column = outputColumns_[i];
+    search::ast::literal_type_bitmask_t literalType;
+    switch (column.type) {
+      case ColumnType::Array:
+        literalType = search::ast::LiteralType::ArrayT;
+        break;
+      case ColumnType::Boolean:
+        literalType = search::ast::LiteralType::BooleanT;
+        break;
+      case ColumnType::Float:
+        literalType = search::ast::LiteralType::FloatT;
+        break;
+      case ColumnType::Integer:
+        literalType = search::ast::LiteralType::IntegerT;
+        break;
+      case ColumnType::String:
+        literalType = search::ast::LiteralType::VarStringT;
+        break;
+      case ColumnType::Timestamp:
+        // TODO: IR timestamp support pending; constrain to Unknown to avoid
+        // mismatched projections.
+        literalType = search::ast::LiteralType::EpochDateT;
+        break;
+      default:
+        literalType = search::ast::LiteralType::UnknownT;
+        break;
+    }
+    result.emplace_back(column.name, literalType);
+  }
+  return result;
+}
+
+ystdlib::error_handling::Result<void> ClpIrCursor::deserialize(
+    uint64_t numRows) {
+  irDeserializer_->get_ir_unit_handler().clearFilteredLogEvents();
+  uint64_t cnt{0};
+  while (cnt < numRows) {
+    auto deserializeResult =
+        irDeserializer_->deserialize_next_ir_unit(*irReader_);
+    if (deserializeResult.has_error()) {
+      auto error = deserializeResult.error();
+      if (std::errc::result_out_of_range == error ||
+          irDeserializer_->is_stream_completed()) {
+        break;
+      }
+      return error;
+    }
+    cnt++;
+  }
+  return ystdlib::error_handling::success();
+}
+
+VectorPtr ClpIrCursor::createVectorHelper(
+    memory::MemoryPool* pool,
+    const TypePtr& vectorType,
+    size_t vectorSize) {
+  if (vectorType->kind() == TypeKind::ROW) {
+    std::vector<VectorPtr> children;
+    auto& rowType = vectorType->as<TypeKind::ROW>();
+    for (uint32_t i = 0; i < rowType.size(); ++i) {
+      children.push_back(
+          createVectorHelper(pool, rowType.childAt(i), vectorSize));
+    }
+    return std::make_shared<RowVector>(
+        pool, vectorType, nullptr, vectorSize, std::move(children));
+  }
+  auto vector = BaseVector::create(vectorType, vectorSize, pool);
+  vector->setNulls(allocateNulls(vectorSize, pool, bits::kNull));
+  VELOX_CHECK_LT(
+      readerIndex_,
+      projectedColumnNameNodeIdMap_.size(),
```

Comment:
```suggestion
      outputColumns_.size(),
```
Otherwise we would crash in situations where not all columns have been resolved yet when we should just be marshalling nulls for unresolved fields.